modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C

```diff
@@ -0,0 +1,101 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdaptiveImportanceStats.h"
+#include "Sampler.h"
+#include "Distribution.h"
+#include "Normal.h"
+#include "libmesh/utility.h"
+
+registerMooseObject("StochasticToolsApp", AdaptiveImportanceStats);
+
+InputParameters
+AdaptiveImportanceStats::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Reporter to compute statistics corresponding to the "
+                             "AdaptiveImportanceSampler.");
+  params.addRequiredParam<ReporterName>("output_value",
+                                        "Value of the model output from the SubApp.");
+  params.addParam<ReporterValueName>("mu_imp", "mu_imp", "Means of the importance distributions.");
+  params.addParam<ReporterValueName>(
+      "std_imp", "std_imp", "Standard deviations of the importance distributions.");
+  params.addParam<ReporterValueName>("pf", "pf", "Failure probability estimate.");
+  params.addParam<ReporterValueName>(
+      "cov_pf", "cov_pf", "Coefficient of variation of failure probability.");
+  params.addRequiredParam<SamplerName>("sampler", "The sampler object.");
+  return params;
+}
+
+AdaptiveImportanceStats::AdaptiveImportanceStats(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _output_value(getReporterValue<std::vector<Real>>("output_value", REPORTER_MODE_DISTRIBUTED)),
+    _mu_imp(declareValue<std::vector<Real>>("mu_imp")),
+    _std_imp(declareValue<std::vector<Real>>("std_imp")),
+    _pf(declareValue<std::vector<Real>>("pf")),
+    _cov_pf(declareValue<std::vector<Real>>("cov_pf")),
+    _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
+    _sampler(getSampler("sampler")),
+    _ais(dynamic_cast<const AdaptiveImportanceSampler *>(&_sampler)),
```

Comment:
Is this object going to be able to accept different samplers in the future. If not:
```suggestion
    _sampler(getSampler<AdaptiveImportanceSampler>("sampler")),
```

Reply:
In this change, we are not getting `_ais` are we? `_ais` is subsequently used in the code.

Reply:
With this change you are getting a reference to `AdaptiveImportanceSampler`. You just need to change type of `_sampler` in the header file. This will also throw an error if the sampler is not AIS, so you don't need the error below.

Reply:
I have addressed this now.